cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	flock "github.com/theckman/go-flock"
+)
+
+// Version is the program's version number.
+const Version = "0.0.1"
+
+func main() {
+	log.SetFlags(0)
+
+	args := getArgs()
+
+	config, err := NewConfig(args.ConfigFile, args.DatabaseDirectory)
+	if err != nil {
+		log.Fatalf("Error loading configuration file: %+v", err)
+	}
+
+	lock, err := setup(config, args.Verbose)
+	if err != nil {
+		log.Fatalf("Error preparing to update: %+v", err)
+	}
+	defer func() {
+		if err := lock.Unlock(); err != nil {
+			log.Fatalf("Error unlocking lock file: %+v", errors.Wrap(err, "unlocking"))
+		}
+	}()
+
+	if err := run(config, args.Verbose); err != nil {
+		log.Fatalf("Error retrieving updates: %+v", err)
+	}
+}
+
+func setup(
+	config *Config,
+	verbose bool,
+) (*flock.Flock, error) {
+	if err := maybeSetProxy(config); err != nil {
+		return nil, err
+	}
+
+	if err := checkEnvironment(config); err != nil {
+		return nil, err
+	}
+
+	lock := flock.NewFlock(config.LockFile)
+	ok, err := lock.TryLock()
+	if err != nil {
+		return nil, errors.Wrap(err, "error acquiring a lock")
+	}
+	if !ok {
+		return nil, errors.Errorf("could not acquire lock on %s", config.LockFile)
+	}
+	if verbose {
+		log.Printf("Acquired lock file lock (%s)", config.LockFile)
+	}
+
+	return lock, nil
+}
+
+var client = &http.Client{
+	Timeout: 30 * time.Minute, // Allow for slow downloads
```

Comment:
Was there a timeout previously? Some people may have really slow connections.

Reply:
There was not AFAICT. We did set up TCP KeepAlives if they were available which would get at the same issue I was worried about (hanging forever). 30 minutes seems generous, but maybe not enough as you say...

I think we should definitely have a non-zero timeout or TCP keepalives though. I'll see about using keepalives instead.

Alternatively we could make the timeout configurable, but probably better to avoid adding options unless it's critical.

Reply:
Looks like we already get TCP KeepAlives by default, I was just ignorant of it: https://golang.org/pkg/net/http/#RoundTripper (`DefaultTransport`'s use of `net.Dialer`'s `KeepAlive` field). In that case, we can do without an explicit timeout.

Reply:
The biggest database update distributed by `geoipupdate` is currently about 4.6 GB. The biggest distributed externally is about 1.2 GB.